### DISCUSSION TITLE
Discussion: namespace from directory structure

### DISCUSSION THREAD

**Discussion by alrz** [dotnet/csharplang#2908] [Language Ideas]
Keeping namespaces in sync with directory structure is usually a convention most of codebases follow. I think such feature would make that easier especially when public API contract is not an issue.

A simple syntax such as `namespace;` is needed to hint the compiler to choose a namespace based on directory structure.

Note: https://github.com/dotnet/csharplang/issues/2546 would only decrease nesting, but directory namespaces would make it entirely implicit.

**Comment by MgSam** [comment]
I'm not a fan of having code files have to be aware of the file system. It means moving a file around now becomes a breaking change. I write a lot of TypeScript as well as C#, and the fact that imports in TS/JS rely on file paths for correctness is one of the most painful and poorly designed things about the language, IMO.

**Comment by HaloFour** [comment]
I'm kind of with @MgSam on this one.  Aside from `CallerFilePathAttribute` I don't think that the location of the source impacts compilation at all.

A feature like this would require a companion feature to establish the root path of the project, as not everyone would want the namespaces hanging off of the root project directory.

**Comment by franchyd** [comment]
That being said this could easily be a plugin/setting on an IDE rather than a language feature

**Comment by DavidArno** [comment]
I'm torn on this one. 

I like the fact that R# warns/errors if the namespace doesn't match the file path (VS may do this natively now days; I've not checked). So I like the idea of not even having to specify the namespace, but to have it inferred by the file's location.

But I think @MgSam also raises some valid points against the idea.

So I've upvoted both the proposal and his counter argument. 😄

**Comment by MV10** [comment]
@HaloFour what, you don't look forward to using libraries like this? 🤣 

```
namespace Users.JohnDoe.Source.Repos.Foo.Bar
```

**Reply by qrli** [reply]
Wow, my home folder name is not English and contains space 🤣

**Reply by Richiban** [reply]
This is a good point; with applications namespaces don't matter so much but for libraries I always take very careful control of namespaces! I want to be free to move my code files around without impacting the names of anything as far as the end user is concerned.

**Comment by ronnygunawan** [comment]
Creating a new .cs file already generates namespace based on directory structure. What you need is an analyzer with code fix to apply `Refactor namespace to match directory structure` fix to all files.

**Comment by MV10** [comment]
@ronnygunawan 
> Creating a new .cs file already generates namespace based on directory structure.

The request is to do this without explicitly specifying the namespace in the code file _at all_. The namespace would be _derived_ from the file location only at compile time.

Bugs galore...

**Comment by HaloFour** [comment]
@ronnygunawan 

> Creating a new .cs file already generates namespace based on directory structure. What you need is an analyzer with code fix to apply `Refactor namespace to match directory structure` fix to all files.

I like this idea.  An analyzer can warn that the namespace doesn't match the directory structure given a configured source root, and can offer a fix to update the namespace to match the location of the file.  This would work similarly to how many IDEs deal with the package names of Java files.

**Comment by jagalves** [comment]
Some comments are really going overboard with the source root configuration. There's already a \<RootNameSpace\> configuration in the .csproj and the root can be the project file itself. No need to overcomplicate things. I mean, the new project type already automatically picks up .cs files under its path, why not automatically assign namespaces as well?

In my opinion this is how it should be:

* namespace declaration (and scope) not set: use the project's default + relative path to .csproj as namespace
* namespace declaration present: as it is today, don't break existing code and allow customization

Example:
* Project: c:\\users\\jagalves\\projects\\Stuff\\MyProject.csproj (RootNamespace "MyProject")
* File: c:\\users\\jagalves\\projects\\Stuff\\Utils\\DateUtils.cs

If the file DateUtils.cs doesn't have a namespace{ } declaration, its namespace shall be "MyProject.Utils".

Today most people just do this but keep the redundant namespace declaration in every file.

**Comment by miloush** [comment]
> If the file DateUtils.cs doesn't have a namespace{ } declaration, its namespace shall be "MyProject.Utils".

This would be a breaking change for code that does not specify namespace currently, as the types go into the global namespace.

**Comment by jagalves** [comment]
Hmm, it's global. I can see now why the OP suggested "namespace;" to trigger the feature.

Still think my suggestion should be considered for the next time .csproj gets breaking changes. Whenever it may be. ;)

**Comment by johnathaningle** [comment]
I would love to see this. I'm not sure if it is possible, but it would be nice if the namespace could be an **optional** statement. If no namespace is provided, the compiler will determine the namespace based on directory structure. This would (hopefully) solve the issue of backwards compatibility.

**Comment by miloush** [comment]
@jingle1000 as just noted namespace is already an optional statement

**Comment by ssg** [comment]
This sounds practical at first but I find it more problematic than [file-scoped directives](https://medium.com/@ssg/a-pledge-for-namespace-directives-in-c-15a322504a15?), because:

- Directories would be both used for semantic (aka public facing) grouping and logical (aka internal) grouping at the same time. That means directory structures like Controllers/Models/Views in an ASP.NET Core project would need to be split into separate namespaces. That means we have to litter the source code with multiple using statements for even small, logical groups like these rather than a single line of namespace directive per file. It would become annoyingly verbose and people would be discouraged to layout the source directory to their liking. This is my greatest issue with this proposal that it lumps semantic and logical grouping into a single scheme.

- There are many hardcoded directories in .NET projects which don't imply a namespace like App_Code, App_Data, App_Resources etc. This would cause confusion, problems.

- It would become impossible to deduce if a project uses namespaces or the global namespace by viewing the source code. You have to browse through csproj file, evaluate if macros if needed, just to assess this. It's unnecessarily vague. "Did the author forget the namespace directive or are they using global namespace?". I know that global namespace use is practically nonexistent which brings me to my next point:

- What if they forget to add "use directories as namespaces" option? That means everything would be lumped up in the global space and this may not be easily detectable, creating global namespace pollution and dirty code.

Even Java, which assumed that a file corresponded to one class, never assumed that a namespace would correspond to a directory.

**Comment by michalczerwinski** [comment]
As an author of the duplicated item I would like to fill in some things:
* it's nice to be able to opt in within specific file
* it would be nice to be able to override namespace for folder in VS (F4 for folder, then property Custom Namespace, similar to existing "Custom Tool Namespace" for items)
* having _Imports.cs might be an additional added value for this feature
* I think this might be a better syntax (but might be a matter of taste)
```csharp
namespace default;
```  
And please mind that this feature is available currently for razor files and it's really convenient!

**Comment by ssg** [comment]
I thought I mentioned that before but couldn't find it. There is also the issue of FS/language discrepancy: Namespaces are case-sensitive in C#, while the filesystems of MacOS and Windows are not. They are case-preserving only. So you can have two namespaces in C# called "A" and "a" but you can't have two separate folders like that.

It might also cause other issues, like not being able to correct the casing in a namespace because source control doesn't care about renames that changes only letter casing. Case in point: https://stackoverflow.com/questions/10523849/changing-capitalization-of-filenames-in-git 

I also find migration path difficult as opposed to file-level namespace directives, which can be a simple refactor option in StyleCop.

**Comment by hez2010** [comment]
C# code can be compiled without file using Roslyn API. What if there's no actual .cs file but only C# code text?

**Reply by mormegil-cz** [reply]
Well, what does the `[CallerFilePath]` attribute do in that case?

**Comment by Trayani** [comment]
I'd rather have an option to declare containing namespace without braces and indentation.

**Reply by ssg** [reply]
and you've got it with C# 10 :)

**Comment by Joe4evr** [comment]
@Trayani #137

**Comment by ChristopherHaws** [comment]
If this were possible, it would pair very nicely with https://github.com/dotnet/project-system/issues/5372

In the csproj you could do something like:
```xml
// MyProject/MyProject.csproj
<ItemGroup>
  <Folder Include="Models" Namespace="MyProject" />
</ItemGroup>
```

and files in that directory could exclude the namespace declaration:
```csharp
// MyProject/Models/FooBar.cs
using System;

public class FooBar
{
}
```

In this case, `FooBar` would be in the `MyProject` namespace.

**Comment by miloush** [comment]
@ChristopherHaws

That would be a breaking change as that is already a valid code, as mentioned above in this thread.

I like the idea that namespace would be a property of folder and the tooling would generate new classes in the namespace as specified in the property. If people need files to be moved across folders and change their namespaces accordingly, then this should be apparent from the code.

**Comment by rcdailey** [comment]


**Reply by 333fred** [reply]
To be clear, you have to consciously enable the analyzers that track these details. They're not enabled by default, and we don't have them enabled in the Roslyn codebase.

> Directory structure is irrelevant

This is how namespaces in C# work. Yes, you can choose to make directory structure mean something. But you by no means have to.

**Comment by Zodt** [comment]
It seems to me that the best option should be the following implementation - explicit and implicit(@alrz) namespaces.

For example, we have a class ``SomeEvent``. It is located in the directory ``../Src/Project/Events/SomeEvent.cs``. 

* In case the namespace is not important to us, you can omit it in the file ``../SomeEvent.cs``. 
Then the compiler will generate a namespace by the directory for that file.
``` csharp
namespace Project.Events // implicit namespace
```

* In case the namespace is important to us, especially if we use ``MassTransit``, we just prescribe the namespace as it is now. 
```csharp 
namespace Project.Messages // explicit namespace
```

**Reply by TahirAhmadov** [reply]
That means you are unable to add something to the global namespace. While generally adding types to the global namespace is undesirable, making it outright impossible is a bad idea.

**Comment by rcdailey** [comment]
I know when it comes to developing language features, it's hard to say "keep it simple" because, in my view, it's the very job of this tooling to hide the complications that make KISS possible for folks like me. Best I can do is share my opinion here and hopefully it contributes to the cause.

For my day-to-day coding, I tend to keep my namespaces and file structure identical. As far as my work experience goes, this just feels like the de facto industry standard. More importantly, tooling like Resharper is constantly pointing out namespaces that do not match my directory structure. So, I've built up a habit over the years of fixing these "issues".

Having to explicitly and redundantly specify namespaces is a hinderance because each time I move a file, I have to "refactor" the namespace, which affects the file itself as well as anywhere that uses it.

If I take a few steps back, I'm not even sure why this convention exists. I come from mainly a C++ background and in that domain, the best practice was to always give a name to namespaces that represents the logical grouping or package of the code. I rarely let directories directly influence the namespace for at least the reason above: When a source file moves, there's no reason for the code to "break".

Maybe it's counter-productive to focus on the language here and instead ask ourselves why tooling like Resharper is doing this and why the convention matters. I'm a bit torn on this, to be honest. However, the conventions I must follow are (most of the time) not something I can change. So I do think it would be helpful for the language to assist in this convention, if it's reasonable to do so. And I only suggest that because of how widely this convention seems to be used; I'm very doubtful this is a fringe issue.

**Reply by HaloFour** [reply]
I almost think it was inherited from Java, where the package name must match the directory structure as that is how the compiler finds sources and classes on the class path.  But I agree, the two should be independent, and for as far as individuals want to maintain that grouping it should be a function of tooling to ensure that the namespace and directory names are kept in sync.

**Reply by pinkfloydx33** [reply]
For what it's worth, you can disable the resharper warnings by disabling the "namespace provider" property in the resharper property pane when clicking on folders in solution explorer. I'm not sure how you do it globally however. Unfortunately I discovered that years after getting used to enforcing the rules and dealing with the refactoring disaster of moving a file. I just can't shake it as the norm now.

**Comment by maxtoroq** [comment]
Another syntax suggestion:

```csharp
implicit namespace;
```

**Comment by VWilcox2000** [comment]
That appeals to me because I almost NEVER want directory / namespace to vary.  Currently visual studio has a nice feature to automatically change namespaces when you drag files around -- if it could show the namespace to side of that, it would work for almost every file.  Not sure if the practice is allowed, but if the compiler would freshly update implicit namespace; to implicit namespace MyProgram.Weather.Forecasting; based on folder and then if I move it elsewhere even outside of visual studio, on compile the namespace updated to the new folder (or if that's a no-no a warning generated that namespace does not match), I'd probably use it.  I would also probably leverage such an option as a project level option for all namespaces if it existed.

**Reply by MV10** [reply]
As someone who thinks directory-based namespaces are a terrible idea, that approach would be nice because undoing it would be as simple as deleting the `implicit` keyword. Globally. _Forever._ 😂

**Reply by VWilcox2000** [reply]
lol, I can see new setup form for Visual Studio:

[  ] Support directory-based namespaces   [   ] Prevent mention of directory-based namespaces
[  ] Use tabs for formatting                         [   ] Prevent tabs making it anywhere near your code.

**Comment by hez2010** [comment]
I don't think this is a good idea: it makes source rely on filesystem and directory structure, which is not always present or reliable.

**Comment by Rekkonnect** [comment]
Here's my take: this is **not** terrible an idea exactly because it would be opt-in. Moving files around during an API change should not be a breaking change for many, and in most cases you want to be structuring your project minimizing the risk of eventually getting it adjusted. You may specify the namespace declaration implication only on files that are the least likely to be moved elsewhere, or not at all. Both VS and Rider already provide a refactoring that automatically changes the namespace, and also apply it on files whose namespaces are already matching the folder structure.

For me personally, I rarely move files around after having pushed a release. Having something like `implicit namespace;` would definitely be convenient for me.

Also, the IDEs could support hinting the resulting namespace for the current placement of the source file in the project.

**Comment by binki** [comment]
Since global usings are already a compiler switch, I’d prefer this to be a compiler switch instead of a new directive in each file. Otherwise, you’re leaving behind unnecessary boilerplate.

**Reply by spydacarnage** [reply]
Global usings aren't a compiler switch - are you thinking about implicit usings?  In which case, activating the switch essentially just adds a source-generated file to your compilation - it doesn't change the meaning of existing code.

**Reply by binki** [reply]
> the switch essentially just adds a source-generated file to your compilation - it doesn't change the meaning of existing code.

With the file-scoped `namespace «identifier»;`  statement, this is all that would be needed.

**Reply by spydacarnage** [reply]
But you'd need to be adding that to each file, which isn't very user-friendly.

**Reply by binki** [reply]
Ah, I didn’t understand what you meant. I thought you were saying a file was implicitly prefixed to every single compilation unit. I researched how global usings and implicit usings work, so I understand your response now. Implicit usings are implemented using global usings and source files generated by the buildsystem itself.

But the very feature I want is for the [IDE0130](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0130) analyzer’s output to be used automatically. The analyzer’s implementation and even the way that embedded resource names are generated by the compiler means that the way to determine a good namespace for a file based on its filename is already solved. It’d be nice if C# could be handled more like `.resx` files in this respect.

**Comment by Phyyl** [comment]
I would love to see this as an opt-in project setting (e.g. `<DirectoryBasedNamespaces>enable</DirectoryBasedNamespaces>`). For some projects with rigorous structure, this would prevent namespaces not matching the folder structure. If you don't like it, don't enable it! Like @ssg mentioned, we would need a convention or an other setting for controlling the generated name... Personally, pascal case should be by default (ignoring spaces).

**Reply by binki** [reply]
> we would need a convention or an other setting for controlling the generated name

Could you expand on this? Why aren’t your folder names matching the actual namespace component names?

**Reply by Phyyl** [reply]
Folders can have spaces, that's the only problem I can see... (ours never have spaces, but I guess this case has to be handled, maybe?)

**Reply by binki** [reply]
> Folders can have spaces, that's the only problem I can see... (ours never have spaces, but I guess this case has to be handled, maybe?)

Ah. Visual Studio already uses an underscore (`_`) in this situation for generated type names. Though making that a compilation error when this feature is turned on is IMO the right approach.

**Reply by HaloFour** [reply]
That's more of a reason to keep the notion of the directory structure out of the language and compiler.  FIlesystems can support all kinds of characters in directory names that would be illegal in a C# identifier.

**Reply by Phyyl** [reply]
The main reason why this came up is that we want to enforce the namespace to match the folder, the namespace is changed when we move files in VS, and we get suggestions (IDE013) when they don't match. It just seems like the next logical step would be to allow devs to opt into something like that for simplicity and convenience.

**Reply by binki** [reply]
Right, if not matching the directory structure can be considered a compiler error, the `namespace` directive itself becomes redundant.

**Reply by HaloFour** [reply]
Tooling could handle that for you without involving the language or the compiler.  I believe that the `namespace` directive should always explicitly be there, regardless of whether or not you prefer the namespace to match the directory structure.  The compiler and language are both entirely agnostic of directory structure, or that the C# source is contained in files at all.

**Reply by binki** [reply]
> The compiler and language are both entirely agnostic of directory structure, or that the C# source is contained in files at all.

How do [`#line`](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/preprocessor-directives#error-and-warning-information) and [caller information](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/caller-information) work then? And, even without that, the compiler is aware of distinct “compilation units” which are the same concept as files minus path information.

> I believe that the `namespace` directive should always explicitly be there

I think `namespace` is redundant boilerplate which is just noise and should ideally not exist. The files belonging to a specific `namespace` should be implied in the same way that files belonging to a specific assembly are implied—namely, by being in the correct directory. This would allow elimination of redundant noise.

I understand that you’re suggesting that we should do away with actually organizing code into directories or use meaningful filenames and instead just use an editor which displays shows a tree based on namespaces. Well, that doesn’t exactly work because the filename itself is still significant (especially for compilation units which contain multiple types) and using multiple directories is necessary to prevent filenames from colliding.

**Reply by HaloFour** [reply]
> How do [#line](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/preprocessor-directives#error-and-warning-information) and [caller information](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/caller-information) work then?

The former works fine with a string of source.  The latter does depend on file information fed by the compiler but it is also entirely optional.

> I think namespace is redundant boilerplate which is just noise and should ideally not exist. The files belonging to a specific namespace should be implied in the same way that files belonging to a specific assembly are implied—namely, by being in the correct directory. This would allow elimination of redundant noise.

I don't care how other people want to organize their projects.  I think putting the source files into directories by namespaces is the redundant bit, not the namespace.  It's a convention borrowed unnecessarily from other languages where source does need to be resolved this way.  But if that's what you want to do, that doesn't bother me.  However, I think making the namespace implicit is detrimental.  I believe you should be able to look at the source by itself, without knowledge of any file structure, and understand where that code fits into the project as a whole.

As for making this optional per project, I think that ripcord should only be pulled in the case of absolute emergencies (like NRTs) and I don't think this remotely rises to the level of justifying a dialect.

I mostly write Java code, where file name and directory structure absolutely do matter and are required to align with the package and class name, and this is something I very rarely think about because the IDE (in this case IntelliJ) will automatically ask me if I want to change the package and class name.  And I like that it asks first, because sometimes I make mistakes.

**Reply by spydacarnage** [reply]
> I think `namespace` is redundant boilerplate which is just noise and should ideally not exist.

It's a single line and helps organise code into meaningful bundles.  I wouldn't consider that redundant...

For instance, if I'm writing a helper assembly, I sometimes like to organise my enums in one folder, extension method classes in another, etc, so they're easier to find.  I wouldn't want my library to have three namespaces to use the base classes, enums, and extension methods at the same time - I want them all to have the same as the parent "folder".

**Reply by MV10** [reply]
> I think `namespace` is redundant boilerplate which is just noise and should ideally not exist.

No offense, but if you work with anything even a little bit complex, the namespace is an essential organizational tool. These days I primarily debug massive codebases written by others (sometimes 10 or 15 years ago where the dev is long gone). We don't typically have the source (or useful things like PDBs, comments, or variable names) -- we work off DLLs extracted from memory dumps. Most of this would be nearly impossible to reasonably review without namespaces grouping large-scale concerns. 

I personally think tying it to any file-system hierarchy is stupid. They're different things with different purposes. For example, in complex systems I make heavy use of subdirectories to group related _files_ which are still part of the same logical namespace. 

Consider an example of where it would just be a pointless hassle. In one of my projects, I decided to treat this simple custom interface as if it was part of the MS DI namespace. Should I have artificially dug three directories deep for a couple lines of code?

```csharp
﻿namespace Microsoft.Extensions.DependencyInjection;

internal interface IAsyncSingleton
{
    public Task InitializeAsync();
}
```

But as long as it can be disabled as my default, I couldn't care less. (But the ILSpy maintainers are going to _hate_ you if they suddenly have to start screwing around with directory structures... and who knows how many decades it'll be before windbg `!wmodule -saveto` would get updated? 😂)

**Reply by Phyyl** [reply]
The idea isn't to remove the namespaces at all or affect existing codebases, just adding an optional feature (opt-in) to generate the namespace based on the folder structure like many languages are doing (e.g. Rust). I understand that from a compiler standpoint it might not make a lot of sense but from a developer standpoint, files and folder structure is how we organize large projects (I work in a large project with tens of thousands of source files). Tools like ILSpy would never see the difference since there is a namespace after all, it's just generated.

**Comment by iam3yal** [comment]
I really think that it's something that the IDE should handle and there's no need to have this coupling between the file system and namespaces even though that generally I like to have them in sync.

p.s. Next someone might suggest implicit types names based on files names. 😄

**Comment by jeswin** [comment]
> I really think that it's something that the IDE should handle and there's no need to have this coupling between the file system and namespaces even though that generally I like to have them in sync.

The emphasis on the "IDE" is what's holding C# back, and relegating it to a mostly MS/Windows thing. Which is a shame considering it's a great language, and now with NativeAOT an alternative to golang. But for adoption, it needs to be usable with just a plain text editor. Which means that unnecessary verbosity needs to be eliminated.

Most modern and growing languages (Python, JS/TS, Rust, Go etc) allow modules/packages/namespaces to be defined via path structure. Directories have been the most natural way to organize files on a computer, and while I understand the need for an escape hatch, we could save some typing by allowing path to namespace equivalence. Especially when a significant number of projects already have them in sync. 

> I'm not a fan of having code files have to be aware of the file system. It means moving a file around now becomes a breaking change.

There are some well-upvoted comments saying this is not how they do things. Sure, but that's just a personal preference. Directories are already a way to organize functions, explicit namespaces to me are redundant. 

And while we're at it, it'll be great to have implicit class names based on the FileName(.cs). :rocket:

**Comment by geekley** [comment]
Since `global using` is a thing, I wish there was a `global namespace` too, so that you don't have to specify it on every file. Or maybe it would be clearer to call it `default namespace` since it would apply only to that folder and subfolders, not to the whole project.

Then all files in that folder and sub-directories would be under that namespace by default, unless it specifies a different one for the file, or another subfolder has another file with a different `default namespace` definition. Plus you could have a corresponding compiler flag and a `<DefaultNamespace>` element in csproj, plus something like `<DefaultNamespacesMatchingFolderStructure>` that automatically assigns these to every subpath in the project.
- Would not break any existing code.
- Would be somewhat consistent with global using.
- Except it would allow complete freedom to manage your implied namespaces with whatever folder structure you want. Because a folder may or may not define a namespace for everything under it - it's up to you.
- The point is so you don't have to write the namespace everywhere making the code simpler.
- Single place for refactoring if you change the namespace.
- If you want every folder to automatically have it matching the folder name, it could be a MsBuild feature that sends the necessary flags to the compiler.

Basically, if any file has a `default namespace [NS];` statement, it also applies to all files under that folder until overriden:

`MyProj/_Globals.cs`
```cs
global using System;
global using System.Collections.Generic;
// etc...
default namespace MyCompany.MyProj;
```
`MyProj/MyClass.cs`
```cs
// namespace MyCompany.MyProj; // implied
```
`MyProj/MyFolder/OtherClass.cs`
```cs
// namespace MyCompany.MyProj; // implied
```
`MyProj/MyComponent/_Globals.cs`
```cs
default namespace MyCompany.MyProj.MyComponent; // override it under this folder
```
`MyProj/MyComponent/Etc/File.cs`
```cs
// namespace MyCompany.MyProj.MyComponent; // implied
```
`MyProj/ResetNS/_Globals.cs`
```cs
default namespace; // resets it under this folder
```
`MyProj/ResetNS/MyFile.cs`
```cs
// default namespace not implied (i.e. it's global::)
```

**Reply by Zodt** [reply]
> Then all files in that folder and sub-directories would be under that namespace by default, unless it specifies a different one for the file, or another subfolder has another file with a different `default namespace` definition. Plus you could have a corresponding compiler flag and a `<DefaultNamespace>` element in csproj, plus something like `<DefaultNamespacesMatchingFolderStructure>` that automatically assigns these to every subpath in the project.

Overall, the approach is interesting, but I can't figure out why to introduce a new tag `DefaultNamespacesMatchingFolderStructure` if `RootNamespace` already exists?

**Reply by geekley** [reply]
@Zodt
`<DefaultNamespacesMatchingFolderStructure>true</...>` in csproj would make it so you don't have to add some `_Globals.cs` files on folders - it would be as if every subpath had one which sets the default namespace there to match the folder structure automatically (actually accomplished through compiler flags). So, this feature request.

Without the tag, you keep existing behavior where you can put code in subfolders freely with no namespace changes (which allows this proposal to not be a breaking change), unless you explicitly use the proposed `default namespace` syntax, which allows you to freely manage implied namespaces in more advanced cases where not every folder necessarily introduces a namespace level. Note how in the example above, MyFolder doesn't introduce a namespace level, but MyComponent does. You could even have a subfolder with an entirely unrelated namespace to the root if you wanted.

I'm suggesting a `<DefaultNamespace>MyCompany.MyProj</DefaultNamespace>` to automatically add this implied namespace at the top level folder, as an alternative to using a file (like how `<Using>` is an alternative to `global using`). The difference is that this sets just the top-level as specified, while the other one sets just the subfolders, according to the folder name.

The existing `<RootNamespace>` tag does something else I think, involving resources? Not sure what, but it doesn't set any implied namespace, you're currently still required to declare it on every file even with this tag. We could alternatively reuse it e.g. by adding an attribute to it instead of introducing `<DefaultNamespace>`, or make `<DefaultNamespace>true</...>` a boolean and reuse the name from `<RootNamespace>`.

**Reply by Zodt** [reply]
@geekley to be honest, I can't think of a single case where the `MyFolder` functionality could be useful. Except in cases where they want to separate the mediator entities by namespace. But this example is incorrect, because with such a separation, a specific namespace subdirectory is usually introduced (note `MyCompany.MyProj.Intermediate`). And I'm even not saying anything about the fact that such an approach borders on bad practices.

Could you clarify the need for this functionality? It's just that the need for such functionality is not obvious to me.

**Reply by jeswin** [reply]
@geekley - some good ideas there, but  tbh I'd be ok for now if MyProj/MyFolder/OtherClass.cs implied `namespace MyCompany.MyProj;`.

However, this is not going to happen. To the detriment of C#, a majority in the community favours keeping the ceremony v/s simplicity. C# adopted a number of functional programming idioms and shorthands, but it's a shame that you can't put a function in a file without the verbosity.

Along the way, it added some incomplete features - like top level statements but in a very limited fashion. It would have been a great idea if it went all the way and allowed implicit namespaces and classes everywhere. I think it'll increasingly cede ground to golang and others, even when C# is in many ways a better language and platform (especially with NativeAOT).

**Reply by geekley** [reply]
> I can't think of a single case where the `MyFolder` functionality could be useful.

@Zodt In my current project, I have a lot of (internal) classes, which I want to separate in folders for organization, but I don't want to put them in separate namespaces, as I don't need that. But it would still be great to not have to repeat the namespace statement on every file. My suggestion gives full flexibility - you can use a folder to introduce a namespace if you want, or not introduce one if you don't want to, or override the namespace entirely.

> And I'm even not saying anything about the fact that such an approach borders on bad practices.

I disagree. Basically, what I'm suggesting is in accordance to the C# way: being as flexible as possible, leaving organization up to the devs (which are presumed mature enough to use the feature properly), instead of enforcing any style that may get in the way for any use case (like in Java, where it forces 1 file = 1 class, and 1 folder = 1 package; which is not the C# way, and can be too restrictive).

I believe it's best to not assume or force restrictions in use case. By letting me organize multiple classes in files however I want, C# makes me ***less*** tempted to use classes incorrectly just because I don't want to create a new file for a small functionality (as opposed to Java enforcing it, which would lead me to add unrelated code in the same file/class). Similarly, I'm proposing to expand this principle to folders/namespaces as well. By introducing this feature while not forcing any particular project structure, the language adapts to what you need instead of the opposite, and you're less tempted to keep loosely-related classes on the same folder if you can separate them better, just because you don't need/want to create separate (public) namespaces for these.

**Comment by jeswin** [comment]
From this thread, it's clear that a part of the community does not want to move towards conciseness; and would rather depend on tools like IDEs to help out with the verbosity and file organization.

I was sufficiently upset with this thread (and the heavy downvotes) that I decided to do something about it. Made TypeScript compile to C#: https://github.com/tsoniclang/tsonic

**Reply by HaloFour** [reply]
> it's clear that a part of the community does not want to move towards conciseness

It's a fine line between "concise" and "terse", and that line is different for everybody.
